src/replay/scheduler.zig

```diff
@@ -0,0 +1,294 @@
+                    },
+                    else => return e,
+                }
+            };
+            assert(try self.thread_pool.trySchedule(self.allocator, .{
+                .transactions = batch.transactions,
+                .results = &self.results,
+                .exit = &self.exit,
+            }));
+            self.batches_started += 1;
+        }
+        return null;
+    }
+};
+
+const ProcessBatchTask = struct {
+    transactions: []const ResolvedTransaction,
+    results: *Channel(?TransactionError),
+    exit: *std.atomic.Value(bool),
+
+    pub fn run(self: *ProcessBatchTask) !void {
+        const result = processBatch(self.transactions, self.exit);
+        if (result != null) self.exit.store(true, .monotonic);
+        try self.results.send(result);
+    }
+};
+
+test "TransactionScheduler: happy path" {
+    const allocator = std.testing.allocator;
+    const Transaction = sig.core.Transaction;
+    var rng = std.Random.DefaultPrng.init(123);
+
+    var thread_pool = ThreadPool.init(.{});
+    var scheduler = try TransactionScheduler.initCapacity(allocator, 10, &thread_pool);
+    defer scheduler.deinit();
+
+    const transactions = [_]Transaction{
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+    };
+    defer for (transactions) |tx| tx.deinit(allocator);
+
+    {
+        const batch1 = try replay.resolve_lookup
+            .resolveBatch(allocator, undefined, transactions[0..3]);
+        errdefer batch1.deinit(allocator);
+
+        const batch2 = try replay.resolve_lookup
+            .resolveBatch(allocator, undefined, transactions[3..6]);
+        errdefer batch2.deinit(allocator);
+
+        scheduler.addBatchAssumeCapacity(batch1);
+        scheduler.addBatchAssumeCapacity(batch2);
+
+        // should be no failures on account collision with the first time this batch was scheduled
+        // scheduler should just know to run it separately
+        // scheduler.addBatchAssumeCapacity(batch1);
+        // scheduler.addBatchAssumeCapacity(batch1);
+        // TODO: add this
+    }
+
+    try std.testing.expectEqual(.done, try replay.confirm_slot.testAwait(&scheduler));
+}
+
+test "TransactionScheduler: failed account locks" {
+    const allocator = std.testing.allocator;
+    const Transaction = sig.core.Transaction;
+    var rng = std.Random.DefaultPrng.init(0);
+
+    var thread_pool = ThreadPool.init(.{});
+    var scheduler = try TransactionScheduler.initCapacity(allocator, 10, &thread_pool);
+    defer scheduler.deinit();
+
+    const tx = try Transaction.initRandom(allocator, rng.random());
+    defer tx.deinit(allocator);
+
+    const unresolved_batch = .{ tx, tx };
```

Comment:
Give this an array result location,
```suggestion
    const unresolved_batch: [2]Transaction = .{ tx, tx };
```
or just
```suggestion
    const unresolved_batch: []const Transaction = &.{ tx, tx };
```

Reply:
0d27f8b7bc7baa41f9bdc67b643eef299e5dddcb